File: modules/camera-zoom/android/src/main/java/expo/modules/camerazoom/CameraZoomModule.kt
package expo.modules.camerazoom

import android.content.Context
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class CameraZoomModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("CameraZoom")

    // Returns the back camera's maximum zoom *ratio* (>= 1.0). expo-camera maps its
    // 0–1 `zoom` prop onto this range, so the JS side converts a desired factor F to
    // zoom = (F - 1) / (max - 1).
    AsyncFunction("getMaxZoomFactor") {
      val context = appContext.reactContext ?: return@AsyncFunction 1.0
      val manager =
        context.getSystemService(Context.CAMERA_SERVICE) as? CameraManager
          ?: return@AsyncFunction 1.0

      var maxZoom = 1.0

      try {
        for (id in manager.cameraIdList) {
          val characteristics = manager.getCameraCharacteristics(id)
          val facing = characteristics.get(CameraCharacteristics.LENS_FACING)
          if (facing != CameraCharacteristics.LENS_FACING_BACK) {
            continue
          }

          // Android 11+ exposes a true zoom-ratio range (includes optical lenses).
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val range = characteristics.get(CameraCharacteristics.CONTROL_ZOOM_RATIO_RANGE)
            if (range != null) {
              maxZoom = range.upper.toDouble()
            }
          }

          // Fallback to the legacy max digital zoom on older devices.
          if (maxZoom <= 1.0) {
            val maxDigital =
              characteristics.get(CameraCharacteristics.SCALER_AVAILABLE_MAX_DIGITAL_ZOOM)
            if (maxDigital != null && maxDigital > 1f) {
              maxZoom = maxDigital.toDouble()
            }
          }
          break
        }
      } catch (e: Exception) {
        return@AsyncFunction 1.0
      }

      maxZoom
    }
  }
}

File: modules/camera-zoom/ios/CameraZoomModule.swift
import AVFoundation
import ExpoModulesCore

public class CameraZoomModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CameraZoom")

    // Returns the back camera's maximum video zoom *factor* (>= 1.0).
    // expo-camera maps its 0–1 `zoom` prop linearly onto 1.0...maxAvailableVideoZoomFactor,
    // so the JS side can convert a desired factor F to zoom = (F - 1) / (max - 1).
    AsyncFunction("getMaxZoomFactor") { () -> Double in
      let discovery = AVCaptureDevice.DiscoverySession(
        deviceTypes: [
          .builtInWideAngleCamera,
          .builtInDualCamera,
          .builtInDualWideCamera,
          .builtInTripleCamera,
        ],
        mediaType: .video,
        position: .back
      )

      let device =
        discovery.devices.first
        ?? AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back)

      guard let device else {
        return 1.0
      }

      return Double(device.maxAvailableVideoZoomFactor)
    }
  }
}
